### ISSUE TITLE
3 Nodes Mesh Network

### ISSUE THREAD

**Issue by ahmadm08** [royyandzakiy/LoRa-RHMesh#2]
Hi @royyandzakiy,

for my final year project, i'm going to implement lora mesh network for power monitoring application

i've spent some time googling for relevant project regarding lora mesh network, there are Meshtastic, Loramesher, and RadioHead RHMesh, and several other libraries. and by chance, i found this project. Thanks a lot

supposed there is node A, node B, and node C

node A act as a gateway between lora mesh network and WiFi, it will push recieved data from node B and C to cloud database.
node B will be put between node A and C, and i expect it will relay message from node C to A, it will also send simulated sensor data to node A.
node C will be connected to another MCU (arduino mega 2560 pro) which is used for data acquisition from the actual sensor. Sensor data will be sent using Serial Comm, and once it recieved, it will immediately send data over mesh network to node A.

the sensor data from node B and C will be send periodically on JSON format just so it would be easier to parse

Below is the scheme of this project

Example 1
<img width="801" alt="Screen Shot 2023-04-01 at 23 36 51" src="https://user-images.githubusercontent.com/63568441/229299173-ee5b5761-1e45-4ef1-a39b-5b9d48237b70.png">

i also want to try it on star topology to test the flexibility of the mesh network

Example 2
<img width="725" alt="Screen Shot 2023-04-01 at 23 37 50" src="https://user-images.githubusercontent.com/63568441/229299187-1eba33d4-1d7f-4993-885e-5dca2f8a4bf6.png">

for the nodes, im gonna use
- esp32
- rfm95w
- OLED screen

these are some of my questions
1. what happened if node a and b send data at the same time, is this RHMesh library reliable for this case and will guarantee that no data will be lost during transmission
2. how can i change or rather force the routing configuration like in example 1 and 2 
3. apart from only sending data periodically from node B and C to node A, i also want to add feature where node A can request data from each the node whenever needed, to test if the communication can be done both ways (bidirectional). What function, method or maybe callback that is very important, that i could use to handle incoming message, and send message at a time or periodically

Thank you in advance

**Comment by royyandzakiy** [comment]
Hi, thanks for the stop. really sorry hadn't got time to get back to this until I completely forgot

I actually am indeed using RHMesh, just to clear things up

1. I have dove into this along time ago, but I forgot (I'll check when I get back to my computer). I think a sender requests some kind of ack back, this helps enable possible QOS 1 implementation. yet in the receiver side, I don't think it has a polling system that holds inread messages, it just processes the first heard message

**Comment by royyandzakiy** [comment]
Hi @ahmadm08 , thanks for stopping by. really sorry hadn't got time to get back to this until I completely forgot about it

I actually am indeed using RHMesh, just to clear things up

1. I have dove into this a long time ago, but I forgot. I think a sender requests some kind of ack back, this helps enable possible QOS 1 implementation. yet in the receiver side, I don't think it has a polling system that holds unread messages, it just processes the first received message, unable to handle collisions (let's say one of them has a stronger rssi). you can though try to implement this in the application level (wrap the current RHMesh).
2. yes, you can force the routes/topology. it requires a little bit of hardcoding, you can inspect the code in RHRouter.cpp (line 223-263). It already has some premade topology examples that forces routing a certain way (it does this by dropping/not processing messages that does not comply the path), and the macro "RH_TEST_NETWORK" needs to be defined (before calling #include "RHMesh.h") to activate this forced topology. you can ofcourse add your own code that resembles your desired topology.
```
...
#ifdef RH_TEST_NETWORK
	if (
#if RH_TEST_NETWORK==1
	    // This network looks like 1-2-3-4
	       (_thisAddress == 1 && _from == 2)
	    || (_thisAddress == 2 && (_from == 1 || _from == 3))
	    || (_thisAddress == 3 && (_from == 2 || _from == 4))
	    || (_thisAddress == 4 && _from == 3)
	    
#elif RH_TEST_NETWORK==2
	       // This network looks like 1-2-4
	       //                         | | |
	       //                         --3--
	       (_thisAddress == 1 && (_from == 2 || _from == 3))
	    ||  _thisAddress == 2
	    ||  _thisAddress == 3
	    || (_thisAddress == 4 && (_from == 2 || _from == 3))
...
```

3. I'd refer this to the sender, intermediary, and receiver. I think a good way to answer your question is simply using the node A as a sender, and chosing the desired address. you can just add a logic to purposfully send with node A to node B and C at a given interval.

**Comment by ahmadm08** [comment]
Thank you. I've completed the programming. I added an HMI display to node 1(A) and utilized it as the coordinator in the network to sequentially request data from each sensor node and upload it to InfluxDB database. I used the millis() function to calculate the delay time from sending a sensor data request to a node until receiving the sensor data. I noticed that the delay will be multiplied by the number of hops for the same length of data. 

For the sensor node, it is very useful to utilize both of the core on the ESP32, so i separated the lora mesh network and sensor acquisition processes. Furthermore, I tested the system for  control application, where the coordinator sends a request to turn a relay on or off, the node then sends a reply confirming the execution of the request.
